Added staggered grid view widget

vinothvino42 opened the pull request:
In this PR, I added Staggered Grid View.

```yaml
body:
  StaggeredGrid: 
    styles: { horizontalGap: 5, verticalGap: 7 }
    item-template:
      data: ${getImages.body}
      name: item
      template:
        CustomCard:
          inputs:
            item: ${item}
    onItemTap: |
      //@code
      console.log("Staggered Grid Item TAPPED - " + event.data.selectedItemIndex);
```

<img src="https://github.com/EnsembleUI/ensemble/assets/12869588/4f502931-cbe4-49fb-859d-e7acaa29e93f" width="350" />

vusters commented on lib/widget/staggered_grid.dart:
```diff
@@ -0,0 +1,141 @@
+  Widget buildWidget(BuildContext context) {
+    final items = buildItems();
+
+    return SingleChildScrollView(
```
this will render all children in 1 shot, so it doesn't really scale. is there a better way to render content as you scroll?

vusters commented on lib/widget/staggered_grid.dart:
```diff
@@ -0,0 +1,141 @@
+  }
+
+  int get getRandomMainAxisCount {
+    return Random().nextInt(3) + 2;
```
this will never generate a count of 1. Intentional? Don't you want to do `nextInt(4) + 1`?

vinothvino42 replied:
I intentionally did like that, 1 in the main axis doesn't look good. So I started to randomize it from the 2

vusters commented on lib/widget/staggered_grid.dart:
```diff
@@ -0,0 +1,141 @@
+          final item = items[index];
+
+          return StaggeredGridTile.count(
+            crossAxisCellCount: 2,
```
randomize this too

vusters commented on lib/widget/staggered_grid.dart:
```diff
@@ -0,0 +1,141 @@
+
+    return SingleChildScrollView(
+      child: StaggeredGrid.count(
+        crossAxisCount: 4,
```
expose this to the user, but default to 4 internally if not specified

vusters reviewed (commented):
Please add to schema.

vinothvino42 commented:
@vusters I just changed the flutter staggered grid view and it now uses `itemBuilder` so only the visible grid tile will be called not all at once. Also added `crossAxisCount` setter.

```YAML

View:
  header:
    title: Staggered Grid View
  onLoad:
    invokeAPI:
      name: getPeople

  body:
    StaggeredGrid: 
      styles: { crossAxisCount: 5, horizontalGap: 5, verticalGap: 5, padding: 10 }
      item-template:
        data: ${getPeople.body.results}
        name: person
        template:
          Image:
            styles: { borderRadius: 4, fit: cover }
            source: ${person.picture.medium}
      onItemTap: |
        //@code
        ensemble.debug("StaggeredGrid item Debugged");
        console.log("StaggeredGridItem onItemTap called: Index - " + event.data.selectedItemIndex);

API:
  getPeople:
    uri: https://randomuser.me/api/?results=1023
    method: GET

```

vusters commented on assets/schema/ensemble_schema.json:
```diff
@@ -1080,6 +1080,52 @@
         }
       }
     },
+    "Staggered-Grid-payload": {
```
this is not enough - you need to use it also. Look at any widget e.g. Button

vinothvino42 replied:
@vusters Yup, I updated it. Can you check the new changes?

vusters commented on lib/widget/staggered_grid.dart:
```diff
@@ -82,29 +86,26 @@ class EnsembleStaggeredGridState extends WidgetState<EnsembleStaggeredGrid>
+        final item = items[index];
+
+        return GestureDetector(
+          onTap: () => _onItemTapped(index),
```
I know we do this elsewhere too, but the user won't know whether the index comes from children or item-template and have to do manual calculations. Can we enforce that either 'children' or 'item-template' can be used at a time. Using both will throw a LanguageError.

vusters replied:
I meant you can use either children or item-template, but not both. Error will be throw if both are specified.

vinothvino42 replied:
Okay, I updated it now

vusters commented on lib/widget/staggered_grid.dart:
```diff
@@ -110,6 +112,13 @@ class EnsembleStaggeredGridState extends WidgetState<EnsembleStaggeredGrid>
     // children will be rendered before templated children
     List<Widget> children = [];
+    if (widget._controller.children != null && templatedChildren != null) {
+      throw LanguageError('Use either children or item-template');
```
Use this string '${EnsembleStaggeredGrid.type} accepts either children or item-template', recovery: 'Please pick either children or item-template'

vinothvino42 replied:
Done @vusters

vusters reviewed (approved):
Looks good. Thank you
